xtevent/_eventgenvars.ado

```diff
@@ -1,306 +1,499 @@
-	return local names = `"`names'"'
-	return local komittrend= "`komittrend'"	
-end
-
-
-
-
-
-
-	
-	
\ No newline at end of file
+* _eventgenvars.ado 1.00 Aug 24 2021
+
+version 11.2
+
+cap program drop _eventgenvars
+program define _eventgenvars, rclass
+
+	#d;
+	syntax [anything] [if] [in],
+	panelvar(varname) /* Panel variable */	
+	timevar(varname) /* Time variable */
+	policyvar(varname) /* Policy variable */
+	[
+	lwindow(real 0)
+	rwindow(integer 0) /* Estimation window. Need to set a default, but it has to be based on the dataset */
+	/* since lwindow and rwindow are now optional, a default value must be provided*/
+	norm(numlist) /* Coefficients to normalize */
+	trend(string) /* Lower limit for trend */
+	
+	impute(string) /*impute policyvar. There are three options: */
+	/*nuchange  imputes outer missing values of policyvar without verifying staggered adoption*/
+	/*stag  imputes outer missing values of policyvar verifying staggered adoption*/
+	/*instag  imputes outer and inner missing values verifying staggered adoption*/
+	STatic
+	
+
+	
+	]	
+	;
+	#d cr	
+	
+	tempvar mz mz2 kg
+	
+	marksample touse
+	
+	* mz maximum of policy outside window
+	* kg grouped event time variable, grouping dummies outside window
+	
+	loc z = "`policyvar'"
+
+	if "`static'"==""{
+		* Get span of calendar time
+		qui su `timevar' if `touse', d
+		loc tmin=r(min)
+		loc tmax=r(max)
+		loc tdiff = `tmax'-`tmin'
+		
+		* Using notation from latest draft 
+		
+		* Endpoints are left: G+L+1
+		* Right: M+1
+		
+		* Left window is G+L
+		* Right window is M 
+		
+		* If you have t=-3 -2 -1 0 1 2 3, tdiff=6 then the maximum window size you can have is 2 on each side
+		* If you have t=-3 -2 -1 0 1 2 3 4, tdiff=7 then the maximum window size you can have is 2 left, 3 right
+			
+		
+		* Error check for window outside event time range
+		if abs(`lwindow') > `=abs(`tdiff')-1' | abs(`rwindow') > `=abs(`tdiff')-1' {
+			di as err _n "Window outside event-time range."
+			exit 301
+		}
+		
+		* Error check for trend event time range
+		if "`trend'"!="" {
+			if abs(`trend') > `=abs(`tdiff')-1' {
+				di as err _n "Trend outside time range."
+				exit 301
+			}
+					
+			* Error check for trend outside window
+			if abs(`trend')>abs(`lwindow') {
+				di as err _n "Trend outside window range."
+				exit 301
+			}
+		}
+		
+		* Check for vars named _k
+		cap unab oldkvars : _k*
+		if !_rc {
+			di as err _n "You have variable names with the _k prefix. _k is reserved for internal -xtevent- variables."
+			di as err _n "Please rename these variables before proceeding."
+			exit 110
+		}
+	}
+
+	********************* find first and last observed values *********************
+
+	*find minimum valid time (time where there is a no-missing observation)
+	tempvar zmint zmint2 zminv zminv2 zmaxt zmaxt2 zmaxv zmaxv2
+	cap by `panelvar' (`timevar'): egen `zmint'=min(`timevar') if !missing(`z') & `touse'
+	cap by `panelvar' (`timevar'): egen `zmint2'=min(`zmint')

```

Comment:
`zmint` is not totally populated due to the condition `!missing(z)` and `zmint2` is the same as `zmint`, but with values for all the observations for each unit. I need a totally populated version of `zmint` because, for each unit, it tells me the time of the first observed value, and then, for each unit, I compare each observation's time with this variable to know if they are placed before or after the first observed value. I had to create `zmint2` because I couldn't do the replacement in `zmint` using `by 
 panelvar (timevar): replace zmint=min(zmint)`

Reply:
Thanks @Constantino-Carreto-Romero! This makes sense to me. Please feel free to resolve the thread.